frame_utils.go

```diff
@@ -0,0 +1,75 @@
+package tchannel
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+)
+
+type RecordingFramePool struct {
+	sync.Mutex
+
+	allocations map[*Frame]string
+	badRelease  []string
+}
+
+func NewRecordingFramePool() *RecordingFramePool {
+	return &RecordingFramePool{
+		allocations: make(map[*Frame]string),
+	}
+}
+
+func (p *RecordingFramePool) Get() *Frame {
+	p.Lock()
+	defer p.Unlock()
+
+	frame := NewFrame(MaxFramePayloadSize)
+	p.allocations[frame] = recordStack()
+	return frame
+}
+
+func (p *RecordingFramePool) Release(f *Frame) {
+	// Make sure the payload is not used after this point by clearing the frame.
+	zeroOut(f.Payload)
+	f.Payload = nil
+	zeroOut(f.buffer)
+	f.buffer = nil
+	zeroOut(f.headerBuffer)
+	f.headerBuffer = nil
+	f.Header = FrameHeader{}
+
+	p.Lock()
+	defer p.Unlock()
+
+	if _, ok := p.allocations[f]; !ok {
+		p.badRelease = append(p.badRelease, "bad Release at "+recordStack())
+		return
+	}
+
+	delete(p.allocations, f)
+}
+
+func (p *RecordingFramePool) CheckEmpty() (int, string) {
```

Comment:
what do you think about returning a struct here that separates into:
 - number of unreleased frames
 - bad allocation messages
 - bad releases

or can combine the last 2, and just do `(unreleasedFrames int, badPooling []string)`